File: clients/go/ahp/multi_host_state_mirror.go
package ahp

import (
	"sync"

	"github.com/microsoft/agent-host-protocol/clients/go/ahptypes"
)

// HostedResourceKey identifies a resource on a particular host.
// Used by [MultiHostStateMirror] as the lookup key.
type HostedResourceKey struct {
	HostID string
	URI    ahptypes.URI
}

// MultiHostStateMirror is a thread-safe map of (host, URI) → state
// snapshot. It deliberately exposes only the bare minimum needed to
// drive a UI that observes multiple hosts simultaneously: write
// snapshots in, read them back, drop them when the host or resource
// disappears.
//
// The mirror has no opinion about how snapshots are kept in sync with
// the server — that's the consumer's job, typically by feeding action
// envelopes from a [HostSubscriptionEvent] stream through the matching
// [ApplyActionToRoot] / [ApplyActionToSession] / [ApplyActionToChat] /
// [ApplyActionToTerminal] reducer and re-storing the result.
type MultiHostStateMirror struct {
	mu      sync.RWMutex
	roots   map[string]ahptypes.RootState
	session map[HostedResourceKey]ahptypes.SessionState
	chat    map[HostedResourceKey]ahptypes.ChatState
	term    map[HostedResourceKey]ahptypes.TerminalState
	changes map[HostedResourceKey]ahptypes.ChangesetState
}

// NewMultiHostStateMirror returns an empty mirror.
func NewMultiHostStateMirror() *MultiHostStateMirror {
	return &MultiHostStateMirror{
		roots:   make(map[string]ahptypes.RootState),
		session: make(map[HostedResourceKey]ahptypes.SessionState),
		chat:    make(map[HostedResourceKey]ahptypes.ChatState),
		term:    make(map[HostedResourceKey]ahptypes.TerminalState),
		changes: make(map[HostedResourceKey]ahptypes.ChangesetState),
	}
}

// PutRoot stores host's root snapshot.
func (m *MultiHostStateMirror) PutRoot(hostID string, root ahptypes.RootState) {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.roots[hostID] = root
}

// Root returns the root snapshot for hostID, or (zero, false) if
// none is recorded.
func (m *MultiHostStateMirror) Root(hostID string) (ahptypes.RootState, bool) {
	m.mu.RLock()
	defer m.mu.RUnlock()
	v, ok := m.roots[hostID]
	return v, ok
}

// PutSession stores a session snapshot under (hostID, uri).
func (m *MultiHostStateMirror) PutSession(hostID string, uri ahptypes.URI, s ahptypes.SessionState) {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.session[HostedResourceKey{hostID, uri}] = s
}

// Session returns the session snapshot at (hostID, uri), or
// (zero, false) if none is recorded.
func (m *MultiHostStateMirror) Session(hostID string, uri ahptypes.URI) (ahptypes.SessionState, bool) {
	m.mu.RLock()
	defer m.mu.RUnlock()
	v, ok := m.session[HostedResourceKey{hostID, uri}]
	return v, ok
}

// PutChat stores a chat snapshot under (hostID, uri).
func (m *MultiHostStateMirror) PutChat(hostID string, uri ahptypes.URI, c ahptypes.ChatState) {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.chat[HostedResourceKey{hostID, uri}] = c
}

// Chat returns the chat snapshot at (hostID, uri), or
// (zero, false) if none is recorded.
func (m *MultiHostStateMirror) Chat(hostID string, uri ahptypes.URI) (ahptypes.ChatState, bool) {
	m.mu.RLock()
	defer m.mu.RUnlock()
	v, ok := m.chat[HostedResourceKey{hostID, uri}]
	return v, ok
}

// PutTerminal stores a terminal snapshot under (hostID, uri).
func (m *MultiHostStateMirror) PutTerminal(hostID string, uri ahptypes.URI, t ahptypes.TerminalState) {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.term[HostedResourceKey{hostID, uri}] = t
}

// Terminal returns the terminal snapshot at (hostID, uri), or
// (zero, false) if none is recorded.
func (m *MultiHostStateMirror) Terminal(hostID string, uri ahptypes.URI) (ahptypes.TerminalState, bool) {
	m.mu.RLock()
	defer m.mu.RUnlock()
	v, ok := m.term[HostedResourceKey{hostID, uri}]
	return v, ok
}

// PutChangeset stores a changeset snapshot under (hostID, uri).
func (m *MultiHostStateMirror) PutChangeset(hostID string, uri ahptypes.URI, c ahptypes.ChangesetState) {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.changes[HostedResourceKey{hostID, uri}] = c
}

// Changeset returns the changeset snapshot at (hostID, uri), or
// (zero, false) if none is recorded.
func (m *MultiHostStateMirror) Changeset(hostID string, uri ahptypes.URI) (ahptypes.ChangesetState, bool) {
	m.mu.RLock()
	defer m.mu.RUnlock()
	v, ok := m.changes[HostedResourceKey{hostID, uri}]
	return v, ok
}

// DropHost removes every snapshot belonging to hostID. Use when a
// host is removed from the multi-host registry.
func (m *MultiHostStateMirror) DropHost(hostID string) {
	m.mu.Lock()
	defer m.mu.Unlock()
	delete(m.roots, hostID)
	for k := range m.session {
		if k.HostID == hostID {
			delete(m.session, k)
		}
	}
	for k := range m.chat {
		if k.HostID == hostID {
			delete(m.chat, k)
		}
	}
	for k := range m.term {
		if k.HostID == hostID {
			delete(m.term, k)
		}
	}
	for k := range m.changes {
		if k.HostID == hostID {
			delete(m.changes, k)
		}
	}
}

// DropResource removes the snapshot at (hostID, uri) across every
// resource kind. No-op if no snapshot exists.
func (m *MultiHostStateMirror) DropResource(hostID string, uri ahptypes.URI) {
	m.mu.Lock()
	defer m.mu.Unlock()
	k := HostedResourceKey{hostID, uri}
	delete(m.session, k)
	delete(m.chat, k)
	delete(m.term, k)
	delete(m.changes, k)
}

File: clients/dotnet/src/AgentHostProtocol/Hosts/MultiHostStateMirror.cs
// Thread-safe (hostId, URI) → state-snapshot mirror.
// Port of multi_host_state_mirror.go.
#nullable enable

using System.Collections.Concurrent;

namespace Microsoft.AgentHostProtocol.Hosts;

/// <summary>
/// Thread-safe map of (hostId, URI) → state snapshot. Port of
/// <c>multi_host_state_mirror.go</c>. Writes snapshots in; reads them back;
/// drops them when the host or resource disappears.
/// </summary>
public sealed class MultiHostStateMirror
{
    // Independent per-key snapshots: ConcurrentDictionary gives lock-free
    // reads and fine-grained writes, which is exactly this access pattern.
    // The per-resource maps key by HostedResourceKey (host + URI value type) so a
    // host id and a URI compose into one collision-free key with value equality —
    // no ad-hoc tuple delimiter to confuse with reserved URI characters.
    private readonly ConcurrentDictionary<HostId, RootState> _roots = new();
    private readonly ConcurrentDictionary<HostedResourceKey, SessionState> _sessions = new();
    private readonly ConcurrentDictionary<HostedResourceKey, ChatState> _chats = new();
    private readonly ConcurrentDictionary<HostedResourceKey, TerminalState> _terminals = new();
    private readonly ConcurrentDictionary<HostedResourceKey, ChangesetState> _changesets = new();

    /// <summary>Stores <paramref name="root"/> for <paramref name="hostId"/>.</summary>
    public void PutRoot(HostId hostId, RootState root)
    {
        Guard.ThrowIfNull(hostId, nameof(hostId));
        Guard.ThrowIfNull(root, nameof(root));
        _roots[hostId] = root;
    }

    /// <summary>Returns the root snapshot for <paramref name="hostId"/>, or (default, false) if absent.</summary>
    public (RootState? Value, bool Found) Root(HostId hostId) =>
        _roots.TryGetValue(hostId, out var v) ? (v, true) : (default, false);

    /// <summary>Stores a session snapshot under (hostId, uri).</summary>
    public void PutSession(HostId hostId, string uri, SessionState state)
    {
        Guard.ThrowIfNull(hostId, nameof(hostId));
        Guard.ThrowIfNull(uri, nameof(uri));
        Guard.ThrowIfNull(state, nameof(state));
        _sessions[new HostedResourceKey(hostId, uri)] = state;
    }

    /// <summary>Returns the session snapshot at (hostId, uri), or (default, false) if absent.</summary>
    public (SessionState? Value, bool Found) Session(HostId hostId, string uri) =>
        _sessions.TryGetValue(new HostedResourceKey(hostId, uri), out var v) ? (v, true) : (default, false);

    /// <summary>Stores a chat snapshot under (hostId, uri).</summary>
    public void PutChat(HostId hostId, string uri, ChatState state)
    {
        Guard.ThrowIfNull(hostId, nameof(hostId));
        Guard.ThrowIfNull(uri, nameof(uri));
        Guard.ThrowIfNull(state, nameof(state));
        _chats[new HostedResourceKey(hostId, uri)] = state;
    }

    /// <summary>Returns the chat snapshot at (hostId, uri), or (default, false) if absent.</summary>
    public (ChatState? Value, bool Found) Chat(HostId hostId, string uri) =>
        _chats.TryGetValue(new HostedResourceKey(hostId, uri), out var v) ? (v, true) : (default, false);

    /// <summary>Stores a terminal snapshot under (hostId, uri).</summary>
    public void PutTerminal(HostId hostId, string uri, TerminalState state)
    {
        Guard.ThrowIfNull(hostId, nameof(hostId));
        Guard.ThrowIfNull(uri, nameof(uri));
        Guard.ThrowIfNull(state, nameof(state));
        _terminals[new HostedResourceKey(hostId, uri)] = state;
    }

    /// <summary>Returns the terminal snapshot at (hostId, uri), or (default, false) if absent.</summary>
    public (TerminalState? Value, bool Found) Terminal(HostId hostId, string uri) =>
        _terminals.TryGetValue(new HostedResourceKey(hostId, uri), out var v) ? (v, true) : (default, false);

    /// <summary>Stores a changeset snapshot under (hostId, uri).</summary>
    public void PutChangeset(HostId hostId, string uri, ChangesetState state)
    {
        Guard.ThrowIfNull(hostId, nameof(hostId));
        Guard.ThrowIfNull(uri, nameof(uri));
        Guard.ThrowIfNull(state, nameof(state));
        _changesets[new HostedResourceKey(hostId, uri)] = state;
    }

    /// <summary>Returns the changeset snapshot at (hostId, uri), or (default, false) if absent.</summary>
    public (ChangesetState? Value, bool Found) Changeset(HostId hostId, string uri) =>
        _changesets.TryGetValue(new HostedResourceKey(hostId, uri), out var v) ? (v, true) : (default, false);

    /// <summary>Removes every snapshot belonging to <paramref name="hostId"/>.</summary>
    public void DropHost(HostId hostId)
    {
        _roots.TryRemove(hostId, out _);
        foreach (var k in _sessions.Keys) if (k.HostId.Equals(hostId)) _sessions.TryRemove(k, out _);
        foreach (var k in _chats.Keys) if (k.HostId.Equals(hostId)) _chats.TryRemove(k, out _);
        foreach (var k in _terminals.Keys) if (k.HostId.Equals(hostId)) _terminals.TryRemove(k, out _);
        foreach (var k in _changesets.Keys) if (k.HostId.Equals(hostId)) _changesets.TryRemove(k, out _);
    }

    /// <summary>Removes the snapshot at (hostId, uri) across every resource kind.</summary>
    public void DropResource(HostId hostId, string uri)
    {
        var key = new HostedResourceKey(hostId, uri);
        _sessions.TryRemove(key, out _);
        _chats.TryRemove(key, out _);
        _terminals.TryRemove(key, out _);
        _changesets.TryRemove(key, out _);
    }
}
